In [14]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

model = RandomForestClassifier(
    n_estimators = 100,
    max_depth = None,
    min_samples_split = 2,
    min_samples_leaf = 1,
    max_features = "sqrt",
    random_state = 42
)


model.fit(X_train, y_train)

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)

print("Accuracy:", accuracy)


Accuracy: 0.9


# 🔥 Now move to Experiment 1 — n_estimators

In [15]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

n_estimators_values = [10, 50, 100, 200]

results = []

for n in n_estimators_values:
    model = RandomForestClassifier(
        n_estimators=n,
        max_depth=None,
        min_samples_split=2,
        min_samples_leaf=1,
        max_features="sqrt",
        random_state=42
    )

    model.fit(X_train, y_train)

    y_pred = model.predict(X_test)

    accuracy = accuracy_score(y_test, y_pred)

    results.append((n, accuracy))

for n, accuracy in results:
    print(f"n_estimators={n} → Accuracy={accuracy:.4f}")

n_estimators=10 → Accuracy=0.9667
n_estimators=50 → Accuracy=0.9000
n_estimators=100 → Accuracy=0.9000
n_estimators=200 → Accuracy=0.9000


Observations:

- 10 trees achieved 96.67% accuracy on the current test set.
- 50, 100, and 200 trees achieved 90.00% accuracy.
- Increasing the number of trees did not improve test accuracy in this experiment.
- More trees generally increase computational cost.
- The effect of n_estimators should be evaluated on appropriate validation data rather than assuming that more trees always produce higher accuracy.

# 🔥 Now move to Experiment 2 — max_depth

In [16]:
max_depth_values = [2, 5, 10, None]

results_depth = []

for depth in max_depth_values:
    model = RandomForestClassifier(
        n_estimators=100,
        max_depth=depth,
        min_samples_split=2,
        min_samples_leaf=1,
        max_features="sqrt",
        random_state=42
    )

    model.fit(X_train, y_train)

    y_pred = model.predict(X_test)

    accuracy = accuracy_score(y_test, y_pred)

    results_depth.append((depth, accuracy))

for depth, accuracy in results_depth:
    print(f"max_depth={depth} → Accuracy={accuracy:.4f}")

max_depth=2 → Accuracy=0.9000
max_depth=5 → Accuracy=0.9333
max_depth=10 → Accuracy=0.9000
max_depth=None → Accuracy=0.9000


# 🔥 Now Experiment 3 — Feature Randomness

In [ ]:
max_features_values = ["sqrt", "log2"]

results_features = []

for feature_setting in max_features_values:
    model = RandomForestClassifier(
        n_estimators=100,
        max_depth=None,
        min_samples_split=2,
        min_samples_leaf=1,
        max_features=feature_setting,
        random_state=42
    )

    model.fit(X_train, y_train)

    y_pred = model.predict(X_test)

    accuracy = accuracy_score(y_test, y_pred)

    results_features.append((feature_setting, accuracy))

for feature_setting, accuracy in results_features:
    print(
        f"max_features={feature_setting} "
        f"→ Accuracy={accuracy:.4f}"
    )

max_features=sqrt → Accuracy=0.9000
max_features=log2 → Accuracy=0.9000


# 🔥 NOW: Section 11 — Compare Results

## 11. Compare Results

### Experiment 1 — Number of Trees

The model was tested with 10, 50, 100, and 200 trees.

The accuracy changed across the different values of n_estimators.
Increasing the number of trees did not necessarily increase the test accuracy
on this particular train-test split.

This shows that adding more trees does not guarantee a higher accuracy.

### Experiment 2 — Tree Depth

The model was tested with max_depth values of 2, 5, 10, and None.

A shallow tree can underfit the data, while a very deep tree can become more
complex and may overfit.

In this experiment, max_depth=5 achieved 0.9333 accuracy, while the other
settings achieved lower accuracy on the test set.

### Experiment 3 — Feature Randomness

The model was tested using max_features="sqrt" and max_features="log2".

Both settings produced the same accuracy of 0.9000 in this experiment.

This shows that changing the number of features considered at each split did
not affect the test accuracy for this particular dataset and split.

# 🔥 Section 12 — Final Observations

## 12. Final Observations

Random Forest combines multiple decision trees to make predictions.

The experiments showed that changing Random Forest hyperparameters can affect
model performance.

The number of trees does not always produce higher accuracy when increased.

The maximum tree depth controls the complexity of individual decision trees.
Very shallow trees may underfit, while unrestricted or very complex trees may
overfit.

The max_features parameter controls how many features are considered when
splitting nodes.

The best hyperparameter setting cannot be determined reliably from a single
train-test split. More robust evaluation methods such as cross-validation can
provide a more reliable estimate of model performance.

Overall, these experiments helped demonstrate how Random Forest hyperparameters
influence model behavior and why hyperparameter tuning is an important part of
machine learning.